core/audits/prioritize-lcp-image.js

```diff
@@ -47,42 +48,72 @@ class PrioritizeLcpImage extends Audit {
    *
    * @param {LH.Artifacts.NetworkRequest} request
    * @param {LH.Artifacts.NetworkRequest} mainResource
-   * @param {Array<LH.Gatherer.Simulation.GraphNode>} initiatorPath
+   * @param {InitiatorPath} initiatorPath
    * @return {boolean}
    */
   static shouldPreloadRequest(request, mainResource, initiatorPath) {
-    const mainResourceDepth = mainResource.redirects ? mainResource.redirects.length : 0;
-
     // If it's already preloaded, no need to recommend it.
     if (request.isLinkPreload) return false;
     // It's not a request loaded over the network, don't recommend it.
     if (NetworkRequest.isNonNetworkRequest(request)) return false;
     // It's already discoverable from the main document, don't recommend it.
-    if (initiatorPath.length <= mainResourceDepth) return false;
+    if (initiatorPath.length <= 2) return false;
     // Finally, return whether or not it belongs to the main frame
     return request.frameId === mainResource.frameId;
   }
 
   /**
    * @param {LH.Gatherer.Simulation.GraphNode} graph
    * @param {NetworkRequest} lcpRecord
-   * @return {{lcpNode: LH.Gatherer.Simulation.GraphNetworkNode|undefined, path: Array<LH.Gatherer.Simulation.GraphNetworkNode>|undefined}}
+   * @return {LH.Gatherer.Simulation.GraphNetworkNode|undefined}
    */
   static findLCPNode(graph, lcpRecord) {
-    let lcpNode;
-    let path;
-    graph.traverse((node, traversalPath) => {
-      if (node.type !== 'network') return;
+    for (const {node} of graph.traverseGenerator()) {
+      if (node.type !== 'network') continue;
       if (node.record.requestId === lcpRecord.requestId) {
-        lcpNode = node;
-        path =
-          traversalPath.slice(1).filter(initiator => initiator.type === 'network');
+        return node;
       }
-    });
-    return {
-      lcpNode,
-      path,
-    };
+    }
+  }
+
+  /**
+   * Get the initiator path starting with lcpRecord back to mainResource, inclusive.
+   * @param {NetworkRequest} lcpRecord
+   * @param {NetworkRequest} mainResource
+   * @return {InitiatorPath}
```

Comment:
this adds two `initiatorType`s that aren't in the CDP: `'redirect'` and `'fallbackToMain'`. Happy to bikeshed on the strings, but these are also only in this audit's debugData, so don't really have to consider the global consequences of that. I'm mostly interested in them for debugging and eventual analysis in HTTP Archive data.

- `'redirect'` - requests that came from a redirect [get the initiator of their redirect source](https://github.com/GoogleChrome/lighthouse/blob/621e10c21972e94c15e1be7702eebecf1834def3/core/lib/network-recorder.js#L94-L101). That makes some sense, but it can also make it easy to miss that a redirect happened if you aren't careful to check. This just makes it explicit.
- `'fallbackToMain'` - for when the initiator path is lost, an explicit annotation that information is missing instead of `'other'` or whatever. I have no sense for how common this is and there's not any way to see all the possible causes in Chrome (they're mostly very implicit), so I'm very interested to see this data in HTTP Archive.